cinder/client.go

```diff
@@ -0,0 +1,264 @@
+	return createSnapshot(c.handleRequest, CreateSnapshotParams{TenantId: c.tenantId, Snapshot: args})
+}
+
+// Lists detailed information for all Block Storage snapshots that the
+// tenant who submits the request can access.
+func (c *Client) GetSnapshotsDetail() (*GetSnapshotsDetailResults, error) {
+	return getSnapshotsDetail(c.handleRequest, GetSnapshotsDetailParams{TenantId: c.tenantId})
+}
+
+// Updates the metadata for a specified snapshot.
+func (c *Client) UpdateSnapshotMetadata(snapshotId, key string) (*UpdateSnapshotMetadataResults, error) {
+	return updateSnapshotMetadata(c.handleRequest, UpdateSnapshotMetadataParams{
+		TenantId:   c.tenantId,
+		SnapshotId: snapshotId,
+		Metadata: UpdateSnapshotMetadataMetadataParams{
+			Key: key,
+		},
+	})
+}
+
+// Creates a volume. To create a bootable volume, include the image
+// ID and set the bootable flag to true in the request body.
+//
+// Preconditions:
+//
+// - The user must have enough volume storage quota remaining to create
+//   a volume of size requested.
+//
+// Asynchronous Postconditions:
+//
+// - With correct permissions, you can see the volume status as
+//   available through API calls.
+// - With correct access, you can see the created volume in the
+//   storage system that OpenStack Block Storage manages.
+//
+// Troubleshooting:
+//
+// - If volume status remains creating or shows another error status,
+//   the request failed. Ensure you meet the preconditions then
+//   investigate the storage backend.
+// - Volume is not created in the storage system which OpenStack Block Storage manages.
+// - The storage node needs enough free storage space to match the
+//   specified size of the volume creation request.
+func (c *Client) CreateVolume(args CreateVolumeVolumeParams) (*CreateVolumeResults, error) {
+	return createVolume(c.handleRequest, CreateVolumeParams{TenantId: c.tenantId, Volume: args})
+}
+
+// Updates a volume.
+func (c *Client) UpdateVolume(volumeId string, args UpdateVolumeVolumeParams) (*UpdateVolumeResults, error) {
+	return updateVolume(c.handleRequest, UpdateVolumeParams{TenantId: c.tenantId, VolumeId: volumeId, Volume: args})
+}
+
+// The volume managed by OpenStack Block Storage is not deleted from the storage system. :
+func (c *Client) DeleteVolume(volumeId string) error {
+	_, err := deleteVolume(
+		c.handleRequest,
+		DeleteVolumeParams{TenantId: c.tenantId, VolumeId: volumeId},
+	)
+	return err
+}
+
+// Lists volume types.
+func (c *Client) GetVolumeTypes() (*GetVolumeTypesResults, error) {
+	return getVolumeTypes(c.handleRequest, GetVolumeTypesParams{TenantId: c.tenantId})
+}
+
+type StatusResultFn func() (string, error)
+
+// VolumeStatusNotifier will check a volume's status to determine
+// whether it matches the given status. After a check, it waits for
+// "waitDur" before attempting again. If the status does not match
+// after "numAttempts", an error is returned.
+func (c *Client) StatusNotifier(
+	getStatus StatusResultFn,
```

Comment:
I suspect the two first params would be better as a single predicate that returns (bool, error). That'll allow for alternative  strategies, like waiting for the status to be in a set, or *not* equal to something, etc.